### PR TITLE
implement HTMLFormControlsCollection and RadioNodeList

### PR DESCRIPTION
**Pull request by UndefinedBehaviour**
https://github.com/jsdom/jsdom/issues/2600

### CODE REVIEW

**Comment by domenic** [comment]
From a first glance, this looks very impressive! I won't have time to do a full review until the weekend, but I am optimistic. Nice work!

**Comment by domenic** [comment]
Do fix the lint though :)

**Comment by UndefinedBehaviour** [comment]
Thank you! lint fixed! 😛

File: lib/jsdom/living/nodes/HTMLFormControlsCollection-impl.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+exports.implementation = class HTMLFormControlsCollectionImpl extends HTMLCollectionImpl {
+  namedItem(key) {
+    if (!key) {
+      return null;
+    }
+    this._update();
+    const nodeList = RadioNodeList.createImpl(this._globalObject, [], {
+      element: this,
+      query: () => this._list.reduce((a, b) => {
+        if (b.getAttributeNS && (b.getAttributeNS(null, "id") === key || b.getAttributeNS(null, "name") === key)) {
```
Comment (domenic) [inline lib/jsdom/living/nodes/HTMLFormControlsCollection-impl.js:15]: When would getAttributeNS not be truthy?

File: lib/jsdom/living/nodes/RadioNodeList-impl.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+"use strict";
+
+const NodeListImpl = require("./NodeList-impl").implementation;
+
+class RadioNodeListImpl extends NodeListImpl {
+  get value() {
+    this._update();
+    const element = this._list.find(e => e._localName === "input" && e.type === "radio" && e.checked);
+    if (element) {
+      return element.value ? element.value : "on";
```
Comment (domenic) [inline lib/jsdom/living/nodes/RadioNodeList-impl.js:10]: Why are you excluding elements with empty string value?

File: lib/jsdom/living/nodes/HTMLFormControlsCollection-impl.js (hunk 1 above)
Comment (UndefinedBehaviour) [reply lib/jsdom/living/nodes/HTMLFormControlsCollection-impl.js]: just looking at the code in HTMLCollection-impl.js I didn't see any obvious check and wanted to play it safe... but yeah I guess if getAttributeNS is missing here there must a bigger problem somewhere else so this is not the place for that check...

File: lib/jsdom/living/nodes/RadioNodeList-impl.js (hunk 2 above)
Comment (UndefinedBehaviour) [reply lib/jsdom/living/nodes/RadioNodeList-impl.js:10]: oops... fixed!
(the wpt suite doesn't check this case 😮)

File: lib/jsdom/living/nodes/RadioNodeList-impl.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+"use strict";
+
+const NodeListImpl = require("./NodeList-impl").implementation;
+
+class RadioNodeListImpl extends NodeListImpl {
+  get value() {
+    this._update();
+    const element = this._list.find(e => e._localName === "input" && e.type === "radio" && e.checked);
```
Comment (domenic) [inline lib/jsdom/living/nodes/RadioNodeList-impl.js:8]: `e.checked` doesn't necessarily represent the checkedness that the spec asks about. As you can see from https://html.spec.whatwg.org/#attr-input-checked it's a pretty complicated relationship. Instead, check `e._checkedness`.

File: lib/jsdom/living/nodes/RadioNodeList-impl.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    if (element && element.value !== "") {
```
Comment (domenic) [inline lib/jsdom/living/nodes/RadioNodeList-impl.js:9]: Let's follow the spec more exactly here. It has a step that says "if element is null, return the empty string". Please translate that, in this case becoming `if (element === undefined) { return ""; }`.

Then it says, "If element is an element with no [value](https://html.spec.whatwg.org/#attr-input-value) attribute, return the string "on"." This is not very well-written spec text, but again, let's try to translate it. In this case it'd be `if (!("value" in element)) { return "on"; }`.

And then there's a final step... that one should be clear.

File: lib/jsdom/living/nodes/RadioNodeList-impl.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      return element.value ? element.value : "on";
+    }
+    return "";
+  }
+  set value(value) {
+    this._update();
```
Comment (domenic) [inline lib/jsdom/living/nodes/RadioNodeList-impl.js:15]: Again, this doesn't follow the spec very well. Please try to translate it more precisely.

File: lib/jsdom/living/nodes/RadioNodeList-impl.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    const element = this._list.find(
+      // https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#dom-radionodelist-value "On setting..."
+      e => e._localName === "input" && e.type === "radio" && (e.value ? value === e.value : value === "on")
+    );
+    if (element) {
+      element.checked = true;
```
Comment (domenic) [inline lib/jsdom/living/nodes/RadioNodeList-impl.js:21]: As above, change the `_checkedness`, not `checked`.

**Review by domenic — CHANGES_REQUESTED** [review]
OK, the full review is here!

In general the main advice is: follow the spec as precisely as possible. The closer of a translation it is, the easier maintenance is in the future, and the easier it is for reviewers or other code-readers to see that it's obviously correct. If you have any questions about how to translate, please feel free to ask them, and I or someone else should be able to chime in pretty fast.

Finally, I suspect a lot of the changes I'm requesting aren't covered by web platform tests. If you have the extra time, it'd be much appreciated to drop extra test cases into the to-upstream web platform tests directory, testing those edge cases. (E.g., cases where `checked` doesn't match the `_checkedness`.) It's not required but would be a nice bonus :).

File: lib/jsdom/living/nodes/RadioNodeList-impl.js (hunk 6 above)
Comment (domenic) [reply lib/jsdom/living/nodes/RadioNodeList-impl.js:21]: Hmm. Actually I suspect the spec is wrong here. Please leave as-is and add a comment linking to https://github.com/whatwg/html/issues/7612 .

If you change this to `_checkedness`, do any tests fail? If none do, then adding one would be super appreciated (but again not required). It would help us straighten out the spec situation.

Comment (UndefinedBehaviour) [reply lib/jsdom/living/nodes/RadioNodeList-impl.js:21]: Yes, if you use `_checkedness` here then a couple tests in `radionodelist.html` will fail as multiple buttons will have their _checkedness set to true

File: lib/jsdom/living/nodes/HTMLFormControlsCollection-impl.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+exports.implementation = class HTMLFormControlsCollectionImpl extends HTMLCollectionImpl {
+  namedItem(key) {
+    if (!key) {
+      return null;
+    }
+    this._update();
+    const nodeList = RadioNodeList.createImpl(this._globalObject, [], {
+      element: this,
+      query: () => this._list.reduce((a, b) => {
```
[the remark this answers is not in the document]
Comment (UndefinedBehaviour) [reply lib/jsdom/living/nodes/HTMLFormControlsCollection-impl.js:14]: Yeah `filter` works and is more straightforward... not sure why i thought i'd use `reduce` here...

**Review by domenic — APPROVED** [review]
